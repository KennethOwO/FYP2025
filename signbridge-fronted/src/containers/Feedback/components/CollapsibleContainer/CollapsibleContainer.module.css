.collapsibleContainer {
    margin-bottom: 10px;
    width: 85%;
}

.collapsibleContent {
    position: relative;
    top: 50%;
    left: 50%;
    transform: translate(-50%, 0);
    width: 100%;
    max-width: 100%;
    margin-bottom: 20px;
    box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
    border-radius: 8px;
    transition: width 0.2s;
    background-color: #fff;
}

.collapsibleContent.notOpened {
    width: 85%;
    cursor: pointer;
}

.collapsibleContent.notOpened:hover {
    width: 100%;
    background-color: rgb(224, 224, 224);
}

.collapsible-content.opened {
    width: 85%;
}

.collapsibleHeader {
    cursor: pointer;
    display: flex;
    justify-content: space-between;
    align-items: center;
    height: 100%;
    padding: 15px;
    background-color: #a4825e;
    border-radius: 8px 8px 0 0;
    font-size: 22px;
    flex-wrap: wrap;
    gap: 10px;
}

.collapsibleHeader h2 {
    font-size: clamp(22px, 2vw, 24px);
    margin: 5px 0;
}

.expandIcon {
    margin-left: 10px;
}

.collapsiblebox {
    margin: 10px 20px 0 20px;
    padding: 10px;
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    gap: 20px;
}

.personalDetails h3,
.ratings h3,
.comments h3 {
    width: 100%;
    background-color: #dccec0;
    padding: 10px;
    margin: 0 auto 10px auto;
    border-radius: 5px;
    font-size: 22px;
    text-align: center;
}

.personalDetails,
.ratings,
.comments {
    flex: 1;
    margin: 0;
    min-width: 300px;
    display: flex;
    flex-direction: column;
    align-items: center;
}

.personalDetailsItem,
.ratingsItem,
.commentsItem {
    margin-left: 10px;
    margin-bottom: 10px;
    font-size: 18px;
    width: 90%;
}

.screenshotImage {
    width: 50%;
    height: auto;
    margin: 10px 0;
}

@media screen and (max-width: 768px) {
    .collapsibleContainer {
        width: 95%;
    }

    .collapsibleContent.notOpened {
        width: 95%;
    }

    .collapsiblebox {
        margin: 10px 10px;
        padding: 5px;
    }

    .personalDetailsItem,
    .ratingsItem,
    .commentsItem {
        font-size: 16px;
    }

    .collapsibleHeader {
        flex-direction: column;
        align-items: flex-start;
        gap: 5px;
    }

    .collapsibleHeader h2 {
        font-size: 18px;
        width: 100%;
        margin: 2px 0;
    }

    .expandIcon {
        position: absolute;
        right: 15px;
        top: 15px;
    }
}

@media screen and (max-width: 700px) {
    .collapsiblebox {
        margin: 10px 10px;
    }
}

@media screen and (max-width: 480px) {
    .collapsibleContainer {
        width: 95%;
    }

    .collapsibleContent.notOpened {
        width: 95%;
    }

    .collapsibleHeader h2 {
        font-size: 18px;
    }

    .collapsiblebox {
        margin: 10px 10px;
    }

    .personalDetails,
    .ratings,
    .comments {
        min-width: 200px;
    }

    .personalDetails h3,
    .ratings h3,
    .comments h3 {
        font-size: 18px;
        padding: 8px;
        width: 90%;
        margin: 0px auto 10px auto;
    }

    .personalDetailsItem,
    .ratingsItem,
    .commentsItem {
        font-size: 14px;
    }
}
