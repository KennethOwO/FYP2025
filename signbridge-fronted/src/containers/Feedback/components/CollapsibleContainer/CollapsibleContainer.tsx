import { useState } from "react";
import style from "./CollapsibleContainer.module.css";
import RatingStars from "../RatingStars/RatingStars";
import { useFeedbackSortFilterStore } from "../../../../store/feedbackSortFilter";
import { useTranslation } from "react-i18next";

interface CollapsibleContainerProps {
  id: number;
  name: string;
  age: number;
  gender: string;
  race: string;
  email: string;
  fcategory: string;
  experience: string;
  friendliness: string;
  quality: string;
  recommended: string;
  q1_en: string;
  q2_en: string;
  q3_en: string;
  q1_bm: string;
  q2_bm: string;
  q3_bm: string;
  image: string;
  created_at: string;
  status_en: string;
  status_bm: string;
  updateStatus: (feedbackId: string, status_en: string, status_bm: string) => void;
}

const CollapsibleContainer: React.FC<CollapsibleContainerProps> = ({
  id,
  name,
  age,
  gender,
  race,
  email,
  fcategory,
  experience,
  friendliness,
  quality,
  recommended,
  q1_en,
  q2_en,
  q3_en,
  q1_bm,
  q2_bm,
  q3_bm,
  image,
  created_at,
  status_en,
  status_bm,
  updateStatus,
}) => {
  const { t, i18n } = useTranslation();
  const [isOpen, setIsOpen] = useState<boolean>(false);
  const useStore = useFeedbackSortFilterStore();

  const toggleOpen = async () => {
    setIsOpen(!isOpen);
    if (!isOpen) {
      updateStatus(id.toString(), "Viewed", "Dilihat");
    }

    for (let i = 0; i < useStore.modifiedData.length; i++) {
      if (useStore.modifiedData[i].feedback_id === id) {
        useStore.setModifiedData([
          ...useStore.modifiedData.slice(0, i),
          {
            ...useStore.modifiedData[i],
            status_en: "Viewed",
            status_bm: "Dilihat",
          },
          ...useStore.modifiedData.slice(i + 1),
        ]);
        break;
      }
    }
  };

  const closeForm = (e: React.MouseEvent<HTMLDivElement>) => {
    const isHeaderClicked =
      e.target === e.currentTarget.querySelector(`.${style.collapsibleHeader}`);

    if (isHeaderClicked) {
      setIsOpen(false);
    }

    e.stopPropagation();
  };

  const personalDetails = [
    { key: "1", label: t("feedback_name"), children: name },
    { key: "2", label: t("age"), children: String(age) },
    { key: "3", label: t("gender"), children: gender },
    { key: "4", label: t("race"), children: race },
    { key: "5", label: t("feedback_email"), children: email },
  ];

  const ratings = [
    { key: "1", label: t("feedback_category_filter"), children: fcategory },
    { key: "2", label: t("view_experience"), children: experience },
    { key: "3", label: t("view_friendliness"), children: friendliness },
    { key: "4", label: t("view_quality"), children: quality },
    { key: "5", label: t("view_recommended"), children: recommended },
  ];

  const comments = [
    { key: "1", label: "Q1", children: i18n.language === "en" ? q1_en : q1_bm },
    { key: "2", label: "Q2", children: i18n.language === "en" ? q2_en : q2_bm },
    { key: "3", label: "Q3", children: i18n.language === "en" ? q3_en : q3_bm },
    { key: "4", label: t("screenshot"), children: image },
  ];

  return (
    <div className={style.collapsibleContainer}>
      <div
        className={`${style.collapsibleContent} ${
          isOpen ? style.opened : style.notOpened
        }`}
        onClick={toggleOpen}
      >
        <div className={style.collapsibleHeader}>
          <h2>
            {t("feedback_id")}: {id}
          </h2>
          <h2>
            {t("feedback_name")}: {name}
          </h2>
          <h2>
            {t("feedback_status")}: {i18n.language === "en" ? status_en : status_bm}
          </h2>
          <h2>
            {t("feedback_date")}: {created_at}
          </h2>
          <div className={style.expandIcon}>
            {isOpen ? <span>&#9650;</span> : <span>&#9660;</span>}
          </div>
        </div>

        {isOpen && (
          <div className={style.collapsibleContent} onClick={closeForm}>
            <div className={style.collapsiblebox}>
              <div className={style.personalDetails}>
                <h3>{t("personal_details")}</h3>
                {personalDetails.map((detail) => (
                  <div key={detail.key} className={style.personalDetailsItem}>
                    <span className={style.personalDetailsLabel}>
                      {detail.label}:{" "}
                    </span>
                    <span className={style.personalDetailsContent}>
                      {detail.children}
                    </span>
                  </div>
                ))}
              </div>
              <div className={style.ratings}>
                <h3>{t("ratings")}</h3>
                {ratings.map((rating) => (
                  <div key={rating.key} className={style.ratingsItem}>
                    <span className={style.ratingsLabel}>
                      {rating.label}
                      {rating.label === t("view_experience") ||
                      rating.label === t("view_friendliness") ||
                      rating.label === t("view_quality") ||
                      rating.label === t("view_recommended")
                        ? ""
                        : ":"}{" "}
                    </span>
                    <span className={style.ratingsContent}>
                      {rating.label === t("view_experience") ||
                      rating.label === t("view_friendliness") ||
                      rating.label === t("view_quality") ||
                      rating.label === t("view_recommended") ? (
                        <RatingStars rating={parseInt(rating.children)} />
                      ) : (
                        rating.children
                      )}
                    </span>
                  </div>
                ))}
              </div>
              <div className={style.comments}>
                <h3>{t("comments")}</h3>
                {comments.map((comment) => (
                  <div key={comment.key} className={style.commentsItem}>
                    <span className={style.commentsLabel}>
                      {comment.label}:{" "}
                    </span>
                    {comment.label === "Screenshot" &&
                    comment.children.toLowerCase().includes("http") ? (
                      <a
                        href={comment.children}
                        target="_blank"
                        rel="noopener noreferrer"
                      >
                        {/* <img src={comment.children} alt="Screenshot" className={style.screenshotImage} /> */}
                        {t("view_screenshot")}
                      </a>
                    ) : (
                      <span className={style.commentsContent}>
                        {comment.children || "-"}
                      </span>
                    )}
                  </div>
                ))}
              </div>
              {/* <div className={style.status}>
                                <h3>Status</h3>
                                <span>{status}</span>
                            </div> */}
            </div>
          </div>
        )}
      </div>
    </div>
  );
};

export default CollapsibleContainer;
