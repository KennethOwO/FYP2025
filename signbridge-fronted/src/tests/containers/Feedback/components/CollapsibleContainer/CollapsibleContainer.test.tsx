import CollapsibleContainer from "@root/containers/Feedback/components/CollapsibleContainer/CollapsibleContainer";
import { create } from "react-test-renderer";
import { cleanup } from "@testing-library/react";
// Mock the translation functions
jest.mock("react-i18next", () => ({
  useTranslation: () => ({
    t: (key: string) => key, // Mock the translation function to return the key
    i18n: {
      changeLanguage: jest.fn(), // Mock the changeLanguage function
    },
  }),
}));

describe("Test CollapsibleContainer", () => {
  afterEach(() => {
    // Clean up the DOM
    cleanup();
    // Reset mocks to their initial state
    jest.resetAllMocks();
  });
  const props = {
    id: 1, // Example number for id
    name: "John Doe", // Example string for name
    age: 30, // Example number for age
    gender: "Male", // Example string for gender
    race: "Human", // Example string for race
    email: "john@example.com", // Example string for email
    fcategory: "Category", // Example string for fcategory
    experience: "Good", // Example string for experience
    friendliness: "Friendly", // Example string for friendliness
    quality: "High", // Example string for quality
    recommended: "Yes", // Example string for recommended
    q1_en: "Question 1 English", // Example string for q1_en
    q2_en: "Question 2 English", // Example string for q2_en
    q3_en: "Question 3 English", // Example string for q3_en
    q1_bm: "Question 1 BM", // Example string for q1_bm
    q2_bm: "Question 2 BM", // Example string for q2_bm
    q3_bm: "Question 3 BM", // Example string for q3_bm
    image: "path/to/image.jpg", // Example string for image path
    created_at: "2024-05-24", // Example string for created_at
    status_en: "Active", // Example string for status_en
    status_bm: "Aktif", // Example string for status_bm
    updateStatus: jest.fn(),
  };
  it("should render correctly", () => {
    const tree = create(<CollapsibleContainer {...props} />);
    expect(tree.toJSON).toMatchSnapshot();
  });
});
