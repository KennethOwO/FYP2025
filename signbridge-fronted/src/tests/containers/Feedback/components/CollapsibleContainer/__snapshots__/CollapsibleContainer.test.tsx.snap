// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CollapsibleContainer should render correctly 1`] = `[Function]`;
